src/vring.rs

```diff
@@ -0,0 +1,253 @@
+// Copyright 2019 Intel Corporation. All Rights Reserved.
+// Copyright 2021 Alibaba Cloud Computing. All rights reserved.
+//
+// SPDX-License-Identifier: Apache-2.0
+
+//! Struct to maintain state information and manipulate vhost-user queues.
+
+use std::fs::File;
+use std::io;
+use std::os::unix::io::{FromRawFd, IntoRawFd};
+use std::result::Result;
+use std::sync::{Arc, RwLock, RwLockReadGuard, RwLockWriteGuard};
+
+use virtio_queue::{Error as VirtQueError, Queue};
+use vm_memory::{GuestAddress, GuestAddressSpace, GuestMemoryAtomic, GuestMemoryMmap};
+use vmm_sys_util::eventfd::EventFd;
+
+/// Struct to maintain raw state information for a vhost-user queue.
+pub struct VringState<M: GuestAddressSpace> {
+    queue: Queue<M>,
+    kick: Option<EventFd>,
+    call: Option<EventFd>,
+    err: Option<EventFd>,
+    enabled: bool,
+}
+
+impl<M: GuestAddressSpace> VringState<M> {
+    fn new(mem: M, max_queue_size: u16) -> Self {
+        VringState {
+            queue: Queue::new(mem, max_queue_size),
+            kick: None,
+            call: None,
+            err: None,
+            enabled: false,
+        }
+    }
+
+    /// Get a mutable reference to the underlying raw `Queue` object.
+    pub fn get_queue_mut(&mut self) -> &mut Queue<M> {
+        &mut self.queue
+    }
+
+    /// Get a immutable reference to the kick event fd.
+    pub fn get_kick(&self) -> &Option<EventFd> {
+        &self.kick
+    }
+}
+
+/// Struct to maintain state information and manipulate a vhost-user queue.
+#[derive(Clone)]
+pub struct Vring<M: GuestAddressSpace = GuestMemoryAtomic<GuestMemoryMmap>> {
+    state: Arc<RwLock<VringState<M>>>,
+}
+
+impl<M: GuestAddressSpace> Vring<M> {
+    /// Get a immutable guard to the underlying raw `VringState` object.
+    pub fn get_ref(&self) -> RwLockReadGuard<VringState<M>> {
+        self.state.read().unwrap()
+    }
+
+    /// Get a mutable guard to the underlying raw `VringState` object.
+    pub fn get_mut(&self) -> RwLockWriteGuard<VringState<M>> {
+        self.state.write().unwrap()
+    }
+
+    /// Add an used descriptor into the used queue.
+    pub fn add_used(&self, desc_index: u16, len: u32) -> Result<(), VirtQueError> {
+        self.get_mut().get_queue_mut().add_used(desc_index, len)
+    }
+
+    /// Notify the vhost-user master that used descriptors have been put into the used queue.
+    pub fn signal_used_queue(&self) -> io::Result<()> {
+        if let Some(call) = self.get_ref().call.as_ref() {
+            call.write(1)
+        } else {
+            Ok(())
+        }
+    }
+
+    /// Enable event notification for queue.
+    pub fn enable_notification(&self) -> Result<bool, VirtQueError> {
+        self.get_mut().get_queue_mut().enable_notification()
+    }
+
+    /// Disable event notification for queue.
+    pub fn disable_notification(&self) -> Result<bool, VirtQueError> {
+        self.get_mut().get_queue_mut().enable_notification()
```

Comment:
This should be a call to `disable_notification`

Reply:
sorry, will fix it:(

Reply:
Fixed